### ISSUE TITLE
self-signing keys are not returned from CS-API for user on different HS

### ISSUE THREAD

**Issue by bwindels** [matrix-org/synapse#7350]
@lampholder is not seeing self-signing keys when calling `/keys/query` on his homeserver for neilj who's on matrix.org. This seems to happen only for this user (neilj), but is 100% reproducable. The master key is the correct one, with `"self_signing_keys": {}` at the end. Tom says he's running `1.12.3`, I've asked him to repro on `1.12.4`, but already wanted to report here as 1.12.4 apparently only contained fixes for workerized instances, which is not the case for Tom.

Exact output returned from the CS-API can be found at https://github.com/matrix-org/riot-web-rageshakes/issues/2658

This causes breakage in Riot as even after verifying somebody, they still appear with a red shield, as trust chain to their devices is incomplete.

**Comment by babolivier** [comment]
Wouldn't that be fixed by https://github.com/matrix-org/synapse/pull/7160? It was also shipped in 1.12.4.

**Comment by bwindels** [comment]
> Wouldn't that be fixed by #7160? It was also shipped in 1.12.4.

Hmm, I don't see the immediate connection between that issue and this, but if you think so. Ideally Tom would retest with 1.12.4 but he is OOO as of today, so don't think he will be updating his synapse soon.

**Comment by anoadragon453** [comment]
Hm, yeah looking at [the list of updates](https://github.com/matrix-org/synapse/compare/release-v1.12.3...release-v1.12.4) I don't think any of the fixes here would help with this.

Looking at my own HS, I get the self-signing keys just fine:

```
{
  "device_keys": {
     ... blabla ...
  },
  "self_signing_keys": {
    "@neilj:matrix.org": {
      "keys": {
        "ed25519:PjWOJEVY2WL6po+pbkw1180RUBCrrGwnEw00fHB5pYc": "PjWOJEVY2WL6po+pbkw1180RUBCrrGwnEw00fHB5pYc"
      },
      "signatures": {
        "@neilj:matrix.org": {
          "ed25519:LL7r5ZWgrQlO5V/Yhus/+fGdv1iOTmyRt0geOMW8PSw": "0ucKZKR7nR4IzNOlg/kllX0gqB/BhoBCw9i1/dRCEe22GSy8jZCgtTsU2aL3AtjuzrvKj1mr02+SjwuiQwvJCw"
        }
      },
      "usage": [
        "self_signing"
      ],
      "user_id": "@neilj:matrix.org"
    }
  },
  "user_signing_keys": {}
}
```

**Comment by bwindels** [comment]
Tom told me he had seen Neil as red since the moment he started using cross-signing. So maybe some federation or other bug was persisted on his HS and never refetched from matrix.org afterwards :man_shrugging:

**Comment by anoadragon453** [comment]
@lampholder Did you get a chance to test on v1.12.4 just in case?

**Comment by bwindels** [comment]
@lampholder is OOO fwiw

**Comment by neilisfragile** [comment]
@bwindels are you aware of any other occurrences in the wild?

**Comment by bwindels** [comment]
I am not, no.

**Comment by flackr** [comment]
I seem to also be running into this. My user @flackr:flack.undo.it has self_signing_keys when queried on my HS and my friend's HS, but not when queried from matrix.org. I set up my keys when cross signing support was first added to https://riot.im/develop. However, other users that have set up self_signing keys more recently (on the now stable cross signing support at https://riot.im/app) have had those keys federated to matrix.org.

**Comment by richvdh** [comment]
ok, as long as this is unreproducible, I don't think there's anything we can do here. matrix.org was struggling yesterday which could explain the earlier problems there.

**Comment by flackr** [comment]
I suspect that if you roll synapse on a homeserver A back to an old version (whichever version was on Matrix.org around the initial cross signing riot release) and then generate cross signing keys on another homeserver B participating in some of the same rooms, that it may reproduce the issue even once A is upgraded.

**Comment by babolivier** [comment]
FTR I believe this issue might have the same cause as https://github.com/matrix-org/synapse/issues/7418

**Comment by flackr** [comment]
I just ran into this again on a new homeserver I created. I am missing everyone's `master_keys` and `self_signing_keys` except for people who have since set up their E2E cross signing keys. I added some details to #7418 in case it is the same issue however I figured I'd also comment here since having recently set up the homeserver I am pretty sure it could be reproduced.

During the initial setup, at one point I had not noticed that my dns provider had implicitly created a redirect for the root domain and was claiming that IP, which meant that when I initially connected to a matrix.org channel my .well-known/matrix/server may not have been accessible at the root domain. This may have lead to it never synchronizing.
